File: lab2/BinarySearch.cpp
// Binary Search: Finds an element in sorted data by halving search space. Time Complexity: O(log n).
#include <bits/stdc++.h>
using namespace std;

int BS(vector<int> &arr, int t, int s, int e) {
    if (e < s) return -1;

    int m = s + (e - s) / 2;

    if (arr[m] == t) return m;
    else if (arr[m] < t) return BS(arr, t, m + 1, e);
    else return BS(arr, t, s, m - 1);
}

double getTime(vector<int> arr, int t) {
    auto t1 = chrono::high_resolution_clock::now();

    BS(arr, t, 0, arr.size() - 1);

    auto t2 = chrono::high_resolution_clock::now();

    return chrono::duration<double, milli>(t2 - t1).count();
}

int main() {
    srand(time(0));

    vector<int> values = {10,100,1000,10000,100000,1000000,10000000};

    for (int n : values) {
        double avg = 0;
        vector<int> arr(n);

        for (int i = 0; i < n; i++) arr[i] = rand();

        sort(arr.begin(), arr.end());

        for (int i = 0; i < 1000; i++) {
            avg += getTime(arr, rand());
        }

        cout << avg / 1000 << ",";
    }

    return 0;
}

File: lab2/python/BinarySearch.py
# Binary Search: Finds an element in sorted data by halving search space.
# Time Complexity: O(log n)


def binary_search(arr, target, start=0, end=None):
    """Return the index of target in sorted arr, or -1 if it is absent."""
    if end is None:
        end = len(arr) - 1

    if end < start:
        return -1

    mid = start + (end - start) // 2
    if arr[mid] == target:
        return mid
    if arr[mid] < target:
        return binary_search(arr, target, mid + 1, end)
    return binary_search(arr, target, start, mid - 1)


def solve():
    arr = [3, 8, 12, 19, 27, 31, 44]
    target = 19
    index = binary_search(arr, target)
    print(f"Array: {arr}")
    print(f"Target {target} found at index: {index}")


if __name__ == "__main__":
    solve()
